### PR TITLE
iosrc.URI: Support parsing, encoding UNC paths

### LINKED ISSUE #1927: Can't open a pcap on a "\\"-notation Windows network share
Repro is with GA Brim tagged `v0.20.0`.

First I've created a Network Share on a Windows 2019 VM in Google Cloud and put a pcap in it.

image

Then from my Windows 10 laptop, I access the remote share via the `\\` notation.

image

At this point if I drag & drop the pcap into my Brim app, I get a failure.

image

This seems to be a side-effect of how we do the pcap import by passing the file path to `zqd` and the `\\` notation perhaps not being well-received. For example, if I try the old trick of doing "Map network drive" so that the remote share is accessed via a drive letter:

image

_Now_ I can import successfully from this "Z: drive".

image

@mason-fish originally witnessed this symptom when working in a Windows VM in Parallels on the Mac. Apparently the default behavior in Parallels has the filesystems on the Mac exposed in Windows as network shares with the `\\` notation. I just wanted to repro it using "vanilla" Windows to confirm that what we were seeing was in no way unique to Parallels.

### PR DESCRIPTION
**Pull request by mattnibs**
UNC paths can include support for a network based file path. Adjust
iosrc.ParseURI to keep hostname of a UNC path with a host. If a file uri has
a host, output the path as a UNC path.

Closes #1927

### CODE REVIEW

File: pkg/iosrc/uri.go [hunk 1]
```diff
@@ -25,6 +27,9 @@ func ParseURI(path string) (URI, error) {
 	if u, ok := stdio(path); ok {
 		return u, nil
 	}
+	if strings.HasPrefix(path, uncPrefix) {
+		return parseUNCPath(path)
+	}
```
[the remark this answers is not in the document]
Comment (mattnibs) [reply pkg/iosrc/uri.go:32]: > UNC is unique to Windows, so it probably shouldn't rear its head outside uri_windows.go

Fair I thought I had somewhere online that this was also supported in unix environments but on second look, I cannot find it. Will fix.

> UNC paths can begin with \\ or //, but this only understands \\

👍🏻

[the remark this answers is not in the document]
Comment (mattnibs) [reply pkg/iosrc/uri.go:32]: @nwt given that you could possibly now have a file uri with a hostname, how do you think we should represent file uris with a hostname when uri.Filepath() is called. Perhaps use the POSIX pathname definition [1]?

```
hostname:/directorypath/resource
```

[1] https://en.wikipedia.org/wiki/Path_(computing)#POSIX_pathname_definition

[the remark this answers is not in the document]
Comment (alfred-landrum) [reply pkg/iosrc/uri.go:32]: Let's only allow UNC paths on Windows: I think then the uncPrefix check can move inside `parseBarePath`, parseUNCPath can move into uri_windows.go, and nothing changes for the unix platforms.

File: pkg/iosrc/uri_windows.go [hunk 2]
```diff
@@ -17,7 +25,32 @@ func parseBarePath(path string) (URI, bool, error) {
...
+	// if z is nil then we have just the host name
+	if z == nil {
+		return URI{Scheme: FileScheme, Host: path}, true, nil
+	}
+	u := URI{Scheme: FileScheme, Host: z[0], Path: z[1]}
+	if !strings.HasPrefix(u.Path, "/") {
+		u.Path = "/" + u.Path
+	}
+	return u, true, nil
 }
```
Comment (nwt) [inline pkg/iosrc/uri_windows.go:54]: This implements [RFC 8089's file URI with Authority mapping for UNC strings](https://tools.ietf.org/html/rfc8089#appendix-E.3.1). If we instead use the [file URI with UNC Path mapping](https://tools.ietf.org/html/rfc8089#appendix-E.3.2), we could simplify this file to
```go
package iosrc

import (
	"path/filepath"
)

func parseBarePath(path string) (URI, bool, error) {
	if filepath.VolumeName(path) == "" {
		if scheme, err := getscheme(path); err != nil || scheme != "" {
			return URI{}, false, err
		}
	}
	path, err := filepath.Abs(path)
	if err != nil {
		return URI{}, false, err
	}
	if len(filepath.VolumeName(path)) == 2 {
		// Add leading '/' to paths beginning with a drive letter.
		path = "/" + path
	}
	return URI{Scheme: FileScheme, Path: filepath.ToSlash(path)}, true, nil
}

func (p URI) Filepath() string {
	path := p.Path
	if path[0] == '/' && len(filepath.VolumeName(path[1:])) == 2 {
		// Strip leading '/' from paths beginning with a drive letter.
		path = path[1:]
	}
	return filepath.FromSlash(path)
}
```

**Review by nwt — APPROVED** [review]

File: pkg/iosrc/uri_windows.go (hunk 2 above)
Comment (mattnibs) [reply pkg/iosrc/uri_windows.go:54]: If it works that would be great. This is much simpler. Let me put this in and see if it passes tests.

Comment (mattnibs) [reply pkg/iosrc/uri_windows.go:22]: Works! Let's do this. Nice tip!

**Review by nwt — APPROVED** [review]
:shipit:

**Comment by philrz** [comment]
I tested by building a `zqd.exe`  based on this branch and swapping it in place of the one that ships with GA Brim tagged `v0.21.1`. As shown in the attached video, it's working now where it wasn't before. 👍 

https://user-images.githubusercontent.com/5934157/103821951-14074e80-5024-11eb-8494-02ced14c5665.mp4

**Pull request merged**